Searching with Postgres

LoayGhreeb opened the pull request:
Closes #11823
Closes #11798
Closes #10490


Resolves https://discourse.jabref.org/t/citations-missing-in-action-in-jabref-5-16/4525/11

### Mandatory checks

<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

InAnYan commented:
I have couple of questions:

1. Is the Postgres **server** one for every library?
2. One library - one **DB**?
3. What is the schema of the DB? If you can explain it shortly.
4. Where the Postgres data is stored? (Somwhere local to user, alongside BIB)

I was thinking if I can use this Postgres service as a vector database for my AI features

InAnYan commented on src/main/java/org/jabref/logic/search/indexing/PostgreIndexer.java:
```diff
@@ -0,0 +1,209 @@
+        task.setTitle(Localization.lang("Indexing bib fields for %0", libraryName));
+        int i = 1;
+        long startTime = System.currentTimeMillis();
+        LOGGER.debug("Adding {} entries to index", entries.size());
```
When I was working on my AI features, we used file(s) for each mention of linked files.

Maybe in your case it can be entry(ies)?

subhramit replied:
> Maybe in your case it can be entry(ies)?

This gets a little inconvenient to read, but nitpick - it's debug log anyway so both work, if you want to go with strict uniformity.

koppor replied:
> This gets a little inconvenient to read, but nitpick

We decided for that to enable more easy covering of Arabic... Full Arabic is hard. See https://phrase.com/blog/posts/pluralization/

---


    zero → “لم نزرع أي شجرة حتى الآن”
    one → “لقد زرعنا شجرة ١ حتى الآن”
    two → “لقد زرعنا شجرتين ٢ حتى الآن”
    few → “لقد زرعنا ٣ شجرات حتى الآن”
    many → “لقد زرعنا ١١ شجرة حتى الآن”
    other → “لقد زرعنا ١٠٠ شجرة حتى الآن”

---

InAnYan commented on src/main/java/org/jabref/logic/search/indexing/PostgreIndexer.java:
```diff
@@ -0,0 +1,209 @@
+        HeadlessExecutorService.INSTANCE.executeAndWait(this::closeIndex);
+    }
+
+    private void closeIndex() {
```
How does this method work?

Does it mean that any time I exit JabRef, all indexes are deleted. And when I open JabRef again, I reindex the library again?

koppor replied:
Yes. Currently for the prototype.

Maybe, it was because of unique ids etc.

Maybe, index needs only to be rebuild if .bib file changed on disk during non-usage of JabRef. (e.g., a user has used notepad++ to modify it)

koppor commented on external-libraries.md:
````diff
@@ -699,8 +706,12 @@ License: BSD-3-Clause
 3. (on WSL) `sed 's/[^a-z]*//' < build/reports/project/dependencies.txt | sed "s/\(.*\) .*/\1/" | grep -v "\->" | sort | uniq > build/dependencies-for-external-libraries.txt`
 
 ```text
+ai.djl.huggingface:tokenizers:0.29.0
````
please add all of them above or do not commit this part of the file...

koppor replied:
I fixed it.

koppor commented:
> I have couple of questions:
> 
>     1. Is the Postgres **server** one for every library?

1 for running JabRef

>     2. One library - one **DB**?

One table per library. We kept it very simple for the search.

>     3. What is the schema of the DB? If you can explain it shortly.

Very simple - just one table

                    CREATE TABLE IF NOT EXISTS "%s" (
                        %s TEXT NOT NULL,
                        %s TEXT NOT NULL,
                        %s TEXT,
                        PRIMARY KEY (%s, %s)
                    )
                    """.formatted(tableName,
                    PostgreConstants.ENTRY_ID,
                    PostgreConstants.FIELD_NAME,
                    PostgreConstants.FIELD_VALUE,
                    PostgreConstants.ENTRY_ID,
                    PostgreConstants.FIELD_NAME));

Feel free to create other tables :)

>     4. Where the Postgres data is stored? (Somwhere local to user, alongside BIB)

In the AppData directory. Even "temp". Thus, very fragile ^^. Not sure if one can change it.

Reason for the fragile setting: The solution we are using is intended for **testing**. The java-german channel said: WTF, please use DuckDB. But we could not manage to get DuckDB up and running that fast. Moreover, Postgres has more plugins for speed. And third, we also support Postgres for remote storage.

You could also think of hooking into the remote postgres. Meaning: Advanced features only work with a remote postgres. (Although I like using the local postgres database. Keeps complexity for the user to a minimum. Only 500 MB additional storage is needed...)

InAnYan commented:
@LoayGhreeb, hi! Chris has found a problem with my AI functionality, when a new instance of JabRef is opened when another is running (https://github.com/JabRef/jabref-issue-melting-pot/issues/575, also https://github.com/JabRef/jabref/pull/11844#pullrequestreview-2335402508).

The reason of this problem, is that we used MVStore as an "embedded database", and it doesn't allow to open two MVStores on the same file.

I was wondering, will your search functionality work correctly if there are two or more JabRef instances running?

And if it will work, then how? Will they use the same DB or different?

LoayGhreeb commented:
> I was wondering, will your search functionality work correctly if there are two or more JabRef instances running?
> 
> And if it will work, then how? Will they use the same DB or different?

It will work because each instance creates a new embedded Postgres server with a different port (see [EmbeddedPostgres#detectPort](https://github.com/zonkyio/embedded-postgres/blob/2f830bf48e7fc894bbfc2451212701704bfe6ebc/src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java#L220-L231)).
I think if a local server is used, one option is to create a new database for each instance but will require an identifier for each instance to be used as the database name. Alternatively, the same database can be used, with each library attempting to create a table, if table creation fails, a number can be appended (e.g., libraryName + i) and retry until successful.

I had a similar issue before with Lucene. The bib fields index is stored in memory because it had to be rebuilt each time, so having multiple instances isn't a problem, but the linked files index is stored on disk and will be used in read-only mode if another instance opens the same library. In this case, the index is used only for searching and doesn’t allow updates.

InAnYan commented:
@LoayGhreeb thanks for answer!

Can I ask one more question? Is it possible to save DB information to disk and restore it later?

LoayGhreeb commented:
> Is it possible to save DB information to disk and restore it later?

Yes, you can. You can set the data directory during server setup using [EmbeddedPostgres#setDataDirectory](https://github.com/zonkyio/embedded-postgres/blob/2f830bf48e7fc894bbfc2451212701704bfe6ebc/src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java#L522), and disable data directory cleaning by setting [EmbeddedPostgres#setCleanDataDirectory](https://github.com/zonkyio/embedded-postgres/blob/2f830bf48e7fc894bbfc2451212701704bfe6ebc/src/main/java/io/zonky/test/db/postgres/embedded/EmbeddedPostgres.java#L517) to `false`.

calixtus commented on src/main/java/org/jabref/gui/importer/actions/SearchGroupsMigrationAction.java:
```diff
@@ -24,11 +23,17 @@ public class SearchGroupsMigrationAction implements GUIPostOpenAction {
     // We cannot have this constant in `Version.java` because of recursion errors
     // Thus, we keep it here, because it is (currently) used only in the context of groups migration.
     public static final Version VERSION_6_0_ALPHA = Version.parse("6.0-alpha");
+    public static final Version VERSION_6_0_ALPHA_1 = Version.parse("6.0-alpha_1");
```
there has been no alpha release yet, its all just CI build yet. So i doubt its correct to introduce another ALPHA iteration step here... @koppor ?

LoayGhreeb replied:
We kept the old version and introduced a new one to ask users who tested the alpha version and migrated their search syntax to Lucene to revert to their backup, as there is no migration from Lucene back to the old syntax.

calixtus replied:
ok i see. But we should do this now for our power users but remove / streamline it before the real alpha release or so. Otherweise everything will be overpolluted with these horrible migration heuristics

koppor replied:
Will be streamlined in the beta version. If the format is kept, we will rewrite `6.0-alpha_1`to `6.0-beta` in the beta version and to `6.0` in the final release and drop reading capabilities for `-alpha_1`.

koppor commented on src/main/resources/l10n/JabRef_en.properties:
```diff
@@ -271,6 +271,7 @@ Each\ line\ must\ be\ of\ the\ following\ form\:\ \'tab\:field1;field2;...;field
 
 Search\ groups\ migration\ of\ %0=Search groups migration of %0
 The\ search\ groups\ syntax\ is\ outdated.\ Do\ you\ want\ to\ migrate\ to\ the\ new\ syntax?=The search groups syntax is outdated. Do you want to migrate to the new syntax?
+The\ search\ groups\ syntax\ has\ been\ reverted\ to\ the\ old\ one.\ Please\ use\ the\ backup\ you\ made\ before\ migrating\ to\ 6.0-alpha. = The search groups syntax has been reverted to the old one. Please use the backup you made before migrating to 6.0-alpha.
```
No need for translation. We will remove this after 6.0 release. It is OK to have it untranslated in JabRef.

koppor commented on src/main/java/org/jabref/model/search/query/SearchQuery.java:
```diff
@@ -148,14 +124,12 @@ public EnumSet<SearchFlags> getSearchFlags() {
-        return Arrays.stream(QueryTermExtractor.getTerms(parsedQuery))
-                     .map(WeightedTerm::getTerm)
-                     .toList();
+        return List.of();
```
Is the whole method needed? If it always returns the empty list, it can be removed?

koppor commented on src/main/java/org/jabref/model/entry/field/FieldFactory.java:
```diff
@@ -55,7 +55,7 @@ public static String serializeOrFieldsList(Set<OrFields> fields) {
      * Checks whether the given field contains LaTeX code or something else
      */
     public static boolean isLatexField(Field field) {
-        return Collections.disjoint(field.getProperties(), Set.of(FieldProperty.VERBATIM, FieldProperty.MARKDOWN));
```
Please also add `SINGLE_ENTR_LINK`.

koppor commented on src/main/java/org/jabref/model/entry/BibEntry.java:
```diff
@@ -483,6 +483,7 @@ public Optional<String> getFieldLatexFree(Field field) {
         } else {
             Optional<String> fieldValue = getField(field);
             if (fieldValue.isPresent()) {
+                // TODO: To we need FieldFactory.isLaTeXField(field) here to filter?
```
```suggestion
                // TODO: Do we need FieldFactory.isLaTeXField(field) here to filter?
```

koppor commented on src/main/java/org/jabref/logic/search/SearchPreferences.java:
```diff
@@ -11,6 +11,7 @@
 import javafx.collections.FXCollections;
 import javafx.collections.ObservableSet;
 
+import org.jabref.model.search.SearchDisplayMode;
```
This import is unnecessary?

koppor commented on src/main/java/org/jabref/gui/preferences/table/TableTabViewModel.java:
```diff
@@ -123,7 +123,6 @@ public void setValues() {
 
         availableColumnsProperty.clear();
         availableColumnsProperty.addAll(
-                new MainTableColumnModel(MainTableColumnModel.Type.MATCH_SCORE),
```
Its a bit a pity that we loose this uncertain matches. But OK to loose if we get matches at all 🤣🤣. -- Maybe also worth noting in the report and/or blog post.

calixtus commented:
Well, we did not even release aloha, why are we already talking about beta, if the it will be a feature of the alpha release?

koppor commented:
> Well, we did not even release aloha, why are we already talking about beta, if the it will be a feature of the alpha release?

Because Software Engineering also discusses the future - and imagines the different cases in the future.  😅

xiaojie2008 commented:
@LoayGhreeb @koppor Hello,

I was just assigned to work on Issue #11428: "Auto-Completion in search bar not working until switching tabs" last night. I noticed that the search functionality is currently being refactored (Issue #11803). In this context, should I continue addressing Issue #11428, or will the refactoring resolve the auto-completion problem once completed?

Thank you for your guidance!

LoayGhreeb commented:
> I was just assigned to work on Issue #11428: "Auto-Completion in search bar not working until switching tabs" last night. I noticed that the search functionality is currently being refactored (Issue #11803). In this context, should I continue addressing Issue #11428, or will the refactoring resolve the auto-completion problem once completed?
> 
> Thank you for your guidance!

You can continue working on #11428. This PR won’t affect the auto-completion classes, so you should be able to address the issue without any conflicts.

xiaojie2008 commented:
> > I was just assigned to work on Issue #11428: "Auto-Completion in search bar not working until switching tabs" last night. I noticed that the search functionality is currently being refactored (Issue #11803). In this context, should I continue addressing Issue #11428, or will the refactoring resolve the auto-completion problem once completed?
> > Thank you for your guidance!
> 
> You can continue working on #11428. This PR won’t affect the auto-completion classes, so you should be able to address the issue without any conflicts.

Thank you for your response. I will continue working on #11428 based on JabRef version 5.13 without pulling in the code from #11803, since you mentioned that this PR does not impact the auto-completion functionality. This allows me to focus on addressing the current auto-completion issue without introducing unnecessary complexity or potential conflicts by pulling a draft PR. If there are any updates or if you need me to test anything on the newer version, please let me know.

Siedlerchr commented on src/main/java/org/jabref/gui/entryeditor/SourceTab.java:
```diff
@@ -233,7 +217,12 @@ private void updateCodeArea() {
             try {
-                codeArea.appendText(getSourceString(currentEntry, mode, fieldPreferences));
+                String sourceText = getSourceString(currentEntry, mode, fieldPreferences);
+                LOGGER.info("sourceText length: {}", sourceText.length());
```
make this debug, info is too much noise

LoayGhreeb replied:
I will change it, it was only for debugging

koppor commented on src/test/java/org/jabref/logic/search/query/LuceneQueryParserTest.java:
```diff
@@ -1,40 +0,0 @@
-                // latex
-                Arguments.of("any:preissinger", "\"prei{\\\\ss}inger\""),
-                Arguments.of("any:jesus", "\"jes{\\\\'{u}}s\""),
-                Arguments.of("any:breitenbucher", "\"breitenb{\\\\\\\"{u}}cher\""),
```
We should move these tests to somewhere else. Didn't find that in the new code.

Test cases should use `Lübeck` as test. (City name in Germany).

Lübeck, Lubeck, Luebeck, L\"{u}beck

LoayGhreeb replied:
> We should move these tests to somewhere else. Didn't find that in the new code.

The bib fields queries are tested in [`src/test/java/org/jabref/logic/search/query/SearchQuerySQLConversionTest.java`](https://github.com/JabRef/jabref/blob/postgresql/src/test/java/org/jabref/logic/search/query/SearchQuerySQLConversionTest.java). I will add these tests there.

> Test cases should use `Lübeck` as test. (City name in Germany).
> 
> Lübeck, Lubeck, Luebeck, L"{u}beck

Only searching with `Lübeck` will work; `Lubeck` and `Luebeck` will not find the entry. This normalization should be done during indexing based on the language. The last term, `L"{u}beck`, will not work because the parser throws an exception due to the `"` character.

koppor commented on src/test/java/org/jabref/logic/search/query/SearchQueryExtractorConversionTest.java:
```diff
@@ -0,0 +1,37 @@
+        return Stream.of(
+                Arguments.of(Set.of("term"), "term"),
+                Arguments.of(Set.of("regex.*term"), "regex.*term"),
+                Arguments.of(Set.of("term"), "any = term"),
```
According to https://docs.jabref.org/finding-sorting-and-cleaning-entries/search#pseudo-fields, it is `anyfield`, not just `any`. I think, we shoud keep `anyfield`. At least as synonym.

LoayGhreeb replied:
Both of them are working see https://github.com/JabRef/jabref/blob/c7baf048e4056a292e37e720dd1a481810a2a6c2/src/main/java/org/jabref/logic/search/query/SearchToSqlVisitor.java#L201

koppor commented on src/test/java/org/jabref/logic/search/query/SearchQueryLuceneConversionTest.java:
```diff
@@ -10,64 +12,55 @@
     @MethodSource
-    void transformationRegularExpression(String expected, String query) {
-        String result = SearchQueryConversion.searchToLucene(query);
+    void testSearchConversion(String expected, String searchExpression) {
```
No test names should begin with `test`. - https://docs.openrewrite.org/recipes/java/testing/cleanup/removetestprefix -- I am not sure why OpenRewrite didn't catch this.

LoayGhreeb replied:
`testSearchConversion` begin with `test` 😅

koppor commented on src/test/java/org/jabref/logic/search/query/SearchQueryLuceneConversionTest.java:
```diff
@@ -10,64 +12,55 @@
 
 class SearchQueryLuceneConversionTest {
```
Is this really lucene? Isn't it Postgres?

LoayGhreeb replied:
> Is this really lucene? Isn't it Postgres?

SQL queries tested in [`SearchQuerySQLConversionTest.java`](https://github.com/JabRef/jabref/blob/postgresql/src/test/java/org/jabref/logic/search/query/SearchQuerySQLConversionTest.java).

 Lucene is used for searching in linked files.

koppor commented on src/test/java/org/jabref/logic/search/query/SearchQueryLuceneConversionTest.java:
```diff
@@ -10,64 +12,55 @@
+                Arguments.of("content:term annotations:term", "any == term"),
+                Arguments.of("content:term annotations:term", "any ==! term"),
 
-                Arguments.of("title:chocolate", "title=chocolate"),
```
I am better in reading concrete names. I somehow miss these basic tests in the new code. (image proceccsing below as kept, but not these ones)

koppor commented on src/test/java/org/jabref/logic/search/query/SearchQueryLuceneConversionTest.java:
```diff
@@ -10,64 +12,55 @@
+                Arguments.of("(-content:\"two term\") (-annotations:\"two term\")", "any != \"two terms\""),
+                Arguments.of("+(-content:imag) +(-annotations:process)", "content != image AND annotations != processing"),
 
-                // String with a space
```
You can keep this comment, because it is not that easy to read the intention of these group of tests.

koppor commented on src/test/java/org/jabref/logic/search/query/SearchQueryLuceneConversionTest.java:
```diff
@@ -10,64 +12,55 @@
+                Arguments.of("MatchNoDocsQuery(\"\")", "title = \"image processing\" AND author = smith"),
 
-                // We renamed fields
-                Arguments.of("any:somecontent", "anyfield = somecontent"),
```
This should be kept? -- Not sure about the current state of discussion `any` versus `anyfield`.

LoayGhreeb replied:
> This should be kept? -- Not sure about the current state of discussion `any` versus `anyfield`.

See https://github.com/JabRef/jabref/pull/11803#discussion_r1798130861

koppor commented on src/test/java/org/jabref/logic/search/query/SearchQueryLuceneConversionTest.java:
```diff
@@ -10,64 +12,55 @@
-
-    public static Stream<Arguments> transformationRegularExpression() {
-        return Stream.of(
-                Arguments.of("any:* AND -groups:/.+/", "groups != .+"),
```
Can this test be kept somehouw? I think, it checks for set groups, which is an impotant use case.

LoayGhreeb replied:
> Can this test be kept somehouw? I think, it checks for set groups, which is an impotant use case.

I kept the test, but the query used for Lucene searches in linked files only. The query `groups != .+` will be converted to `org.apache.lucene.search.MatchNoDocsQuery`. The only searchable fields in linked files are `content` and `annotations`; any other fields will be ignored and converted to `MatchNoDocsQuery`. Queries without a specified field or using fields like `any` or `anyfield` will search in both `content` and `annotations`.

koppor commented on src/test/java/org/jabref/logic/search/query/SearchQueryLuceneConversionTest.java:
```diff
@@ -10,64 +12,55 @@
-    public static Stream<Arguments> transformationRegularExpression() {
-        return Stream.of(
-                Arguments.of("any:* AND -groups:/.+/", "groups != .+"),
-                Arguments.of("( any:* AND -groups:/.+/ ) AND ( any:* AND -readstatus:/.+/ )", "groups != .+ and readstatus != .+"),
```
This should also be kept: non-empty group and set read-status

LoayGhreeb replied:
> This should also be kept: non-empty group and set read-status

This is only for searching linked files using Lucene. The searchable fields in linked files are `content` and `annotations`. Searching in bib fields is handled by SQL queries. I will add those test cases to `org.jabref.logic.search.query.SearchQuerySQLConversionTest`.

LoayGhreeb commented on src/main/resources/l10n/JabRef_en.properties:
```diff
@@ -810,14 +808,15 @@ Character\ encoding\ '%0'\ is\ not\ supported.=Character encoding '%0' is not su
 Searching...=Searching...
 Finished\ Searching=Finished Searching
 Search\ expression=Search expression
+This\ only\ affects\ unfielded\ terms.\ For\ using\ RegEx\ in\ a\ fielded\ term,\ use\ \=~\ operator=This only affects unfielded terms. For using RegEx in a fielded term, use =~ operator
+This\ only\ affects\ unfielded\ terms.\ For\ using\ case-sensitive\ in\ a\ fielded\ term,\ use\ \=!\ operator=This only affects unfielded terms. For using case-sensitive in a fielded term, use =! operator
```
```suggestion
This\ only\ affects\ unfielded\ terms.\ For\ using\ RegEx\ in\ a\ fielded\ term,\ use\ \=~\ operator=This only affects unfielded terms. For using RegEx in a fielded term, use =~ operator.
This\ only\ affects\ unfielded\ terms.\ For\ using\ case-sensitive\ in\ a\ fielded\ term,\ use\ \=!\ operator=This only affects unfielded terms. For using case-sensitive in a fielded term, use =! operator.
```

LoayGhreeb commented on src/main/resources/l10n/JabRef_en.properties:
```diff
@@ -844,7 +843,7 @@ Clear\ search=Clear search
 Search\ for\ unlinked\ local\ files=Search for unlinked local files
 Search\ full\ text\ documents\ online=Search full text documents online
-Hint\:\n\nTo\ search\ all\ fields\ for\ <b>Smith</b>,\ enter\:\n<tt>smith</tt>\n\nTo\ search\ the\ field\ <b>author</b>\ for\ <b>Smith</b>\ and\ the\ field\ <b>title</b>\ for\ <b>electrical</b>,\ enter\:\n<tt>author\:Smith\ AND\ title\:electrical</tt>=Hint:\n\nTo search all fields for <b>Smith</b>, enter:\n<tt>smith</tt>\n\nTo search the field <b>author</b> for <b>Smith</b> and the field <b>title ...
+Hint\:\n\nTo\ search\ all\ fields\ for\ <b>Smith</b>,\ enter\:\n<tt>smith</tt>\n\nTo\ search\ the\ field\ <b>author</b>\ for\ <b>Smith</b>\ and\ the\ field\ <b>title</b>\ for\ <b>electrical</b>,\ enter\:\n<tt>author\=\ Smith\ AND\ title\=\ electrical</tt>=Hint:\n\nTo search all fields for <b>Smith</b>, enter:\n<tt>smith</tt>\n\nTo search the field <b>author</b> for <b>Smith</b> and the field <b>t ...
```
```suggestion
Hint\:\n\nTo\ search\ all\ fields\ for\ <b>Smith</b>,\ enter\:\n<tt>smith</tt>\n\nTo\ search\ the\ field\ <b>author</b>\ for\ <b>Smith</b>\ and\ the\ field\ <b>title</b>\ for\ <b>electrical</b>,\ enter\:\n<tt>author\=Smith\ AND\ title\=electrical</tt>=Hint:\n\nTo search all fields for <b>Smith</b>, enter:\n<tt>smith</tt>\n\nTo search the field <b>author</b> for <b>Smith</b> and the field <b>title</b> for <b>electrical</b>, enter:\n<tt>author=Smith AND title=electrical</tt>
```

koppor commented:
Also think about if/when to handle chinese characters (https://github.com/JabRef/jabref/issues/9605). Does a user input pinyin?

koppor commented:
It might be that for our Cli application, we need to fire up postgres on some cases --> follow-up issue (to really think about the cases where we need it and how we could conditionally fire it up)

koppor commented:
I merged main. Should be OK.

However, I cannot search for words separated by space

![image](https://github.com/user-attachments/assets/be351c3d-3080-415f-860d-41416b068f83)

LoayGhreeb commented:
> Also think about if/when to handle chinese characters (#9605). Does a user input pinyin?

Currently, searching with Chinese characters will work. If we want to support pinyin, it will be a follow-up. As discussed before, we want to implement normalization based on language, creating new columns for each language and possibly using Lucene's NormalizationFilter classes or finding other libraries for this purpose.

---

> It might be that for our Cli application, we need to fire up postgres on some cases --> follow-up issue (to really think about the cases where we need it and how we could conditionally fire it up)

Yes, we have [Export matching entries](https://docs.jabref.org/advanced/commandline#export-matching-entries-m-field-searchterm-outputfile-file-exportformat) from the command line. It should index the library and export the matches. Implemented  in `org.jabref.logic.search.DatabaseSearcher`

---

> I merged main. Should be OK.
> 
> However, I cannot search for words separated by space

I see, and the parser throws an exception in that case. I will try to fix this before merging. I also added a comment on the parser test https://github.com/JabRef/jabref/blob/df2e12b6a55bcb360ff09301b99bec2178f5ef02/src/test/java/org/jabref/logic/search/query/SearchQuerySQLConversionTest.java#L661-L662

LoayGhreeb commented on src/main/java/org/jabref/logic/search/query/SearchQueryExtractorVisitor.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Override
+    public Set<String> visitUnaryExpression(SearchParser.UnaryExpressionContext ctx) {
+        isNegated = !isNegated;
```
> I miss a test for this class. Currently, it reads strange. (If the test is src/test/java/org/jabref/logic/search/query/SearchQueryExtractorConversionTest.java, link it in JavaDoc of this class`)

Yes, the test is in [`org.jabref.logic.search.query.SearchQueryExtractorConversionTest`](https://github.com/JabRef/jabref/blob/postgresql/src/test/java/org/jabref/logic/search/query/SearchQueryExtractorConversionTest.java), but I forgot to add the JavaDoc link.

> Maybe, one does not need the variable
> `isNegated` at all. The checks in `visitComparison` read fine.

The `isNegated` variable, is necessary to handle the unary `NOT` operator, it differs from the `!=` operator.

The `isNegated = !isNegated` logic covers cases like `NOT (NOT term)`, where the `term` should be included in the search terms in such cases.

LoayGhreeb commented on src/test/java/org/jabref/logic/search/DatabaseSearcherWithBibFilesTest.java:
```diff
@@ -81,62 +83,58 @@ private BibDatabaseContext initializeDatabaseFromPath(String testFile) throws Ex
-                Arguments.of(List.of(MINIMAL_SENTENCE_CASE, MINIMAL_ALL_UPPER_CASE, MINIMAL_MIXED_CASE), "test-library-with-attached-files.bib", "This is a short sentence, comma included.", EnumSet.of(SearchFlags.FULLTEXT)),
-                Arguments.of(List.of(MINIMAL_SENTENCE_CASE, MINIMAL_ALL_UPPER_CASE, MINIMAL_MIXED_CASE), "test-library-with-attached-files.bib", "comma", EnumSet.of(SearchFlags.FULLTEXT)),
-
-                // TODO: PDF search does not support case sensitive search (yet)
-                // Arguments.of(List.of(minimalAllUpperCase, minimalMixedCase), "test-library-with-attached-files.bib", "THIS", EnumSet.of(SearchFlags.FULLTEXT, SearchFlags.CASE_SENSITIVE)),
-                // Arguments.of(List.of(minimalAllUpperCase), "test-library-with-attached-files.bib", "THIS is a short sentence, comma included.", EnumSet.of(SearchFlags.FULLTEXT, SearchFlags.CASE_SENSITIVE)),
-                // Arguments.of(List.of(minimalSentenceCase, minimalAllUpperCase, minimalMixedCase), "test-library-with-attached-files.bib", "comma", EnumSet.of(SearchFlags.FULLTEXT, SearchFlags.CASE_SENSITIVE)),
-                // Arguments.of(List.of(minimalNoteAllUpperCase), "test-library-with-attached-files.bib", "THIS IS A SHORT SENTENCE, COMMA INCLUDED.", EnumSet.of(SearchFlags.FULLTEXT, SearchFlags.CASE_SENSITIVE)),
```
> This is still a TODO, isn't it? - I didn't see it covered in the added test cases

I think these test cases should be removed, as Lucene doesn't support case-sensitive searches. The tests are for linked files, not bib fields.

LoayGhreeb commented on src/test/java/org/jabref/logic/search/DatabaseSearcherWithBibFilesTest.java:
```diff
@@ -81,62 +83,58 @@ private BibDatabaseContext initializeDatabaseFromPath(String testFile) throws Ex
-                Arguments.of(List.of(MINIMAL_NOTE_SENTENCE_CASE, MINIMAL_NOTE_ALL_UPPER_CASE, MINIMAL_NOTE_MIXED_CASE), "test-library-with-attached-files.bib", "world", EnumSet.of(SearchFlags.FULLTEXT)),
-                Arguments.of(List.of(MINIMAL_NOTE_SENTENCE_CASE, MINIMAL_NOTE_ALL_UPPER_CASE, MINIMAL_NOTE_MIXED_CASE), "test-library-with-attached-files.bib", "Hello World", EnumSet.of(SearchFlags.FULLTEXT))
-
-                // TODO: PDF search does not support case sensitive search (yet)
-                // Arguments.of(List.of(minimalNoteAllUpperCase), "test-library-with-attached-files.bib", "HELLO WORLD", EnumSet.of(SearchFlags.FULLTEXT, SearchFlags.CASE_SENSITIVE)),
-                // Arguments.of(List.of(), "test-library-with-attached-files.bib", "NotExisting", EnumSet.of(SearchFlags.FULLTEXT, SearchFlags.CASE_SENSITIVE))
```
> This is still a TODO, isn't it? - I didn't see it covered in the added test cases

I think these test cases should be removed, as Lucene doesn't support case-sensitive searches. The tests are for linked files, not bib fields.

koppor commented on src/main/java/org/jabref/gui/importer/actions/SearchGroupsMigrationAction.java:
```diff
@@ -24,11 +23,17 @@ public class SearchGroupsMigrationAction implements GUIPostOpenAction {
+            if (currentVersion.get().equals(VERSION_6_0_ALPHA)) {
+                // TODO: This text will only be shown after releasing 6.0-alpha and then removed
+                dialogService.showErrorDialogAndWait("Search groups migration of " + parserResult.getPath().map(Path::toString).orElse(""),
+                        "The search groups syntax has been reverted to the old one. Please use the backup you made before migrating to 6.0-alpha.");
```
```suggestion
                        "The search groups syntax has been reverted to the old one. Please use the backup you made before using 6.0-alpha.");
```

Siedlerchr commented:
Yeah finally!

calixtus commented:
Yay!

koppor commented:
Follow-up to https://github.com/JabRef/jabref/pull/11542
